src/Microsoft.ML.StandardTrainers/Standard/LogisticRegression/MulticlassLogisticRegression.cs

```diff
@@ -992,35 +950,280 @@ DataViewRow ICanGetSummaryAsIRow.GetSummaryIRowOrNull(RoleMappedSchema schema)
+            : base(env, RegistrationName, weights, numClasses, numFeatures, labelNames, stats)
+        {
+        }
+
+        internal LinearMulticlassModelParameters(IHostEnvironment env, VBuffer<float>[] weights, float[] bias, int numClasses, int numFeatures, string[] labelNames, LinearModelStatistics stats = null)
+            : base(env, RegistrationName, weights, bias, numClasses, numFeatures, labelNames, stats)
+        {
+        }
+
+        private LinearMulticlassModelParameters(IHostEnvironment env, ModelLoadContext ctx)
+            : base(env, RegistrationName, ctx)
+        {
+        }
+
+        /// <summary>
+        /// This function does not do any calibration. It's common in multi-class support vector machines where probabilitic outputs are not provided.
+        /// </summary>
+        /// <param name="dst">Score vector should be calibrated.</param>
+        private protected override void Calibrate(Span<float> dst)
+        {
+            return;
+        }
+
+        private static LinearMulticlassModelParameters Create(IHostEnvironment env, ModelLoadContext ctx)
         {
             Contracts.CheckValue(env, nameof(env));
-            var host = env.Register("TrainLRMultiClass");
-            host.CheckValue(input, nameof(input));
-            EntryPointUtils.CheckInputArgs(host, input);
+            env.CheckValue(ctx, nameof(ctx));
+            ctx.CheckAtModel(VersionInfo);
+            return new LinearMulticlassModelParameters(env, ctx);
+        }
 
-            return TrainerEntryPointsUtils.Train<LogisticRegressionMulticlassClassificationTrainer.Options, CommonOutputs.MulticlassClassificationOutput>(host, input,
-                () => new LogisticRegressionMulticlassClassificationTrainer(host, input),
-                () => TrainerEntryPointsUtils.FindColumn(host, input.TrainingData.Schema, input.LabelColumnName),
-                () => TrainerEntryPointsUtils.FindColumn(host, input.TrainingData.Schema, input.ExampleWeightColumnName));
+        /// <summary>
+        /// Output the text model to a given writer.
+        /// </summary>
+        private protected override void SaveAsCodeCore(TextWriter writer, RoleMappedSchema schema)
+        {
+            Host.CheckValue(writer, nameof(writer));
+            Host.CheckValueOrNull(schema);
+
+            for (int i = 0; i < Biases.Length; i++)
+            {
+                LinearPredictorUtils.SaveAsCode(writer,
+                    in Weights[i],
+                    Biases[i],
+                    schema,
+                    "score[" + i.ToString() + "]");
+            }
+
+            writer.WriteLine(string.Format("var softmax = MathUtils.SoftMax(scores.AsSpan(0, {0}));", NumberOfClasses));
+            for (int c = 0; c < Biases.Length; c++)
+                writer.WriteLine("output[{0}] = Math.Exp(scores[{0}] - softmax);", c);
+        }
+
+        private protected override bool SaveAsOnnxCore(OnnxContext ctx, string[] outputs, string featureColumn)
```

Comment:
>SaveAsOnnxCore [](start = 40, length = 14)

It looks like the following three methods are very similar, if not the same, for `LinearMulticlassModelParameters` and `MaximumEntropyModelParameters`.
Can they be refactored in the base class?

Reply:
Done. Thanks.

---
In reply to: [267075643](https://github.com/dotnet/machinelearning/pull/2976#discussion_r267075643) [](ancestors = 267075643)